### DISCUSSION TITLE
How can I load only existing preload data?

### DISCUSSION THREAD

**Discussion by positivedeveloper36** [adonisjs/core#3836] [Help]
I have a relation between my category table and my dishes table but when I preload dishes it loads dishes which have empty relation as well

My code like this :

```
const data = await Category
    .query()
    .preload('dishes',(query) => {
        query.select('id', 'name);
    })
    .paginate(1,2)
```

I want only those category which have relation with dishes

I try use whereHas like this : 
```
const data = await Category
    .query()
    .whereHas('dishes', () => {})
    .preload('dishes',(query)=>{
    query.where('id',name);
   }).paginate(1,2);
```

It works. It success show only those category which have relation with dishes

But is that the best way? Because this uses whereHas and preload together

**Comment by Barbapapazes** [comment]
Hello,

'`request.input('user_id')` doesn't exist' is a valide use case of your application?

If not, you can [try to use a validator](https://docs.adonisjs.com/guides/validator/introduction) to be sure that this input exist!

**Reply by positivedeveloper36** [reply]
@Barbapapazes You don't seem to understand what I mean. I updated my question. Hope you know what I 
mean

**Reply by Barbapapazes** [reply]
> But is that the best way? Because this uses whereHas and preload together

Yes! What do you think is wrong using preload and whereHas? Under the hood, the request is 

```sql
# If dishes have one categories and categories have many dishies
select * from "categories" where exists (select * from "dishes" where "dishies"."id" = ? and "dishes"."categorie_id" = "categories"."id");
# If there is no related categories, the row is not relected in th subquery so the categorie is no selected in the main query.
```
And this is a valide query and I can't thing you can do a better job than this. You can manually use a `join` or a `in` but here, you're main request is not affected (no new column, only data from the categories table will be fetched).

The `preload` is not present in this SQL query because it's done along side the main query.

Hope I anwser!

**Reply by positivedeveloper36** [reply]
@Barbapapazes I use model so I don't use `join` fo query builder. What do yo mean `in`?

**Reply by positivedeveloper36** [reply]
Yes I have checked that. If I debug the query and run it in pgadmin. It looks a bit weird like you said. So to make sure the data match I'm using postman and I think the best way is use whereHas and preload together. Because when I use both and check in postman it displays correct data
